### DISCUSSION TITLE
what do we need from the v2/v3 framework classes?

### DISCUSSION THREAD

**Discussion by markusschloesser** [markusschloesser/MackieC4_P3#95] [Ideas]
collection of things I am finding/thinking about, of which things we actually need to represent current functionality.
So we need 2 sides of things:
1. stuff from Live, like mixer, channel strip, device, etc (Live calls them COMPONENTS, found in ableton\v2\control_surface\components \)
2. stuff from C4, things like (Live calls them CONTROLS, they are found in ableton\v2\control_surface\control\ or in 11.2 even v3...) 
a. RingedEncoderElement with display and pushability X 32, and 
b. special buttons, some with LED ( FUNCTION (Split/Lock, Spot Erase), ASSIGNMENT (Marker, Track, Chan Strip, Function)), some without LEDs: MODIFIERS, PARAMENTER Bank and Single, and SLOT-UP/DOWN, Track left/right

found this on some russian forum:
![HdFCAl-gejE](https://user-images.githubusercontent.com/59286549/194672721-82ac382b-997e-4c5c-94d1-b59b4ec06100.jpg)

which relates to the LOM:
![ObjectModel](https://user-images.githubusercontent.com/59286549/194672751-22226756-6c78-423f-8696-6c6491c09d38.png)

see https://github.com/markusschloesser/MackieC4_P3/discussions/73 for already found stuff last year

**Comment by markusschloesser** [comment]
ChannelStrip component => like current **ChanStrip** mode
DeviceCompontent => like current **Track** Mode (device editing)

Then we have the current **Function** mode, which is kind of a global mode. For this we need TransportComponent, and others (to find out)

For the currently unused **Marker** mode, either do a Mixer with MixerComponent or a clip launching thingie with SessionComponent (could be interesting with showing the names of the clips above the vpot encoder and using the ring like in Live's session view)

**Comment by BilldarBagdar** [comment]
Hi Markus,

I hope rather than starting a new topic, that commenting here is a good enough place to mention this idea.  

I bought a development program yesterday called [Control Surface Studio](https://remotify.io/product/control-surface-studio) (desktop version) based on some of the associated YouTube videos, but mostly because I have a crazy idea about programming a (now freeware) application called [Musicwonk ](http://algoart.com/)to be a kind of "Algorithmic DJ" as well as "Algorithmic keyboard player".  I.E. I want to write a Live "remote script" so Live will treat Musicwonk like a control surface (and of course also make Musicwonk patches that behave like an "Algorithmic DJ" operating itself as both a control surface and a "keyboard player").  

I think CSS will work great helping me realize the above idea.  But I don't think I really know enough about how everything works together in CSS yet to just take off marching to my own drummer on some crazy vector like that.  I'm thinking about first attempting to convert this C4 script into the CSS "format" and continuing to work on it in CSS.  This is the idea I'm mentioning.

Have you heard of Control Surface Studio?  I suspect the v2/v3 framework question posed in this thread would become moot, more or less.

Would you be opposed to such a (radical?) change in direction for our MackieC4 script collaboration project?  Actually, I don't think opposed is the right word because I'm not sure anything would necessarily change here at GitHub.  I don't think CSS produces "source code" artifacts like would normally get archived in a git repo.   

One of the things I like about CSS is the "live object model" is menu driven for script writers, the events are all listed in nested menus so you just pick the one you want.

**Reply by BilldarBagdar** [reply]
I uploaded version 0.0.1 of a Mackie C4 "Control Surface Template" file for use with CSS to their forum.  The way I understand the situation now, it means you or anyone can freely download that template for use with CSS desktop, or sign up for the free account at remotify.io and use that template with either the paid or complimentary online version of CSS.  Either way, you or anyone can keep and use any "scripts" you create using whatever version.  

Aside from the super tight "script development" feedback loop evident in all the videos, I think the CSS product feature that really sealed the deal for me was showcased in the video on "[reactions](https://www.youtube.com/watch?v=iYtotPDoa3k)".  In other words, the topic of the video is "Listeners run code in reaction to being notified when an Event has occurred".  All you need to think about while developing with CSS is how you want your device and Live to interact.  You don't need to think about any of the underlying/supporting python scaffolding/infrastructure unless until you hit a wall, and then "custom python code" options remain available.  

I'm not sure I (we?) will be able to completely replicate all behavior from this script in a CSS C4 script immediately, I mean without additional CSS features being added by the developer over time.  But I am sure most of the functionality can be replicated now, including for example, updating the LCD screens via SYSEX in response to events.  I'm not yet sure if CSS will natively allow, for example, the "master volume" metering this script can display.  I don't know for sure if I can make a CSS script run code in response to an "on display update" Event, but I think I can.  I also think I can simplify what happens when the "on display update" method runs (i.e. when that Listener reacts to that event)

Proof, however, will be in the pudding.  We'll see how far I actually get.

**Reply by markusschloesser** [reply]
Hi, i copied your comments to a new discussion and will "hide" this here, to keep it "clean" as my idea pad 😇☺
https://github.com/markusschloesser/MackieC4_P3/discussions/96

**Comment by markusschloesser** [comment]
also https://github.com/markusschloesser/MackieC4_P3/issues/76
